apps/prairielearn/elements/pl-file-editor/pl-file-editor.js

```diff
@@ -26,6 +26,94 @@ window.PLFileEditor = function (uuid, options) {
   this.editor.setReadOnly(options.readOnly);
   this.editor.getSession().on('change', this.syncFileToHiddenInput.bind(this));
 
+  this.session = this.editor.getSession();
+  this.doc = this.session.getDocument();
+  this.hasRanges = options.hasRanges;
+  this.rangeList = [];
+  
+
+  let editor = this.editor,
+      session = this.session,
+      doc = this.doc;
+
+function before(obj, method, wrapper) {
+  var orig = obj[method];
+  obj[method] = function() {
+    var args = Array.prototype.slice.call(arguments);
+    return wrapper.call(this, function(){
+      return orig.apply(obj, args);
+    }, args);
+  }
+
+  return obj[method];
+}
+
+function intersects(range) {
+  var cursor = editor.selection.getCursor();
+  return (
+    cursor.row >= range.start.row && cursor.row <= range.end.row &&
+    cursor.column >= range.start.column && cursor.column <= range.end.column
+  );
+}
+
+ preventReadonly = (next, args) => {
+  if (this.rangeList.ranges.some(intersects)) return;
+  next();
+}
+before(editor, 'onPaste', preventReadonly);
+before(editor, 'onCut',   preventReadonly);
+
+  applyGutterStyles = () => {
+  var lines = editor.session.getLength();
+  for (var i = 0; i < lines; i++) {
+    if (this.rangeList.some(range =>
+      i >= range.start.row && i <= range.end.row
+    )) {
+      editor.session.addGutterDecoration(i, 'readonly-gutter');
+    } else {
+      editor.session.addGutterDecoration(i, 'editable-gutter');
+    }
+  }
+}
+
+ updateGutter = () => {
+  editor.session.$decorations = [];
+  setTimeout(() => {
+    applyGutterStyles();
+  }, 0);
+}
+
+editor.on("change", updateGutter);
+editor.on("changeSelection", updateGutter);
+
+applyGutterStyles();
+  if (this.hasRanges) {
+    editor.keyBinding.addKeyboardHandler({
```

Comment:
I don't like the idea of doing this with keyboard handlers. This leaves behind changes that involve other methods of input (e.g., right-clicking to paste or cut, dragging text with a mouse, etc). Can this be done with `onchange` instead?

Reply:
None of us that worked on this update are skilled Javascript developers so we stole most of the code from this [Stack Overflow Post](https://stackoverflow.com/questions/39640328/how-to-make-multiple-chunk-of-lines-readonly-in-ace-editor). We'll look into `onchange`. Right now drag and drop and copy and paste work with the editable sections. You also can't copy into the static sections. There is a bug that will allow you to drag and drop code into the static sections.